### ISSUE TITLE
Calling effect cleanup multiple times not working

### ISSUE THREAD

**Issue by sheikalthaf** [angular/angular#58211]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

Calling the effect cleanup multiple times inside the effect function only considers the last cleanup one. I don't know whether it is a bug or missing feature. Till today i thought it is supported similar to `DestroyRef`.

```ts

class TestComponent {
    constructor() {
        effect((onCleanup) => {
            const target = this.target();
            ....
            const resizeObserver = new ResizeObserver(() => {
                ....
            });
            resizeObserver.observe(target());
            onCleanup(() => resizeObserver.disconnect());

            if (checkScroll) {
                target.addEventListener('scroll', this.updateDimension);
                onCleanup(() => target.removeEventListener('scroll', this.updateDimension)); // Only this cleanup gets called
            }
        });
    }
}
```

### Please provide the environment you discovered this bug in (run `ng version`)

Angular CLI: 18.1.1
Node: 20.17.0
Package Manager: pnpm 9.8.0
OS: win32 x64

Angular: 18.1.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, platform-browser, platform-browser-dynamic
... platform-server, router, ssr

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1801.1
@angular-devkit/build-angular   18.1.1
@angular-devkit/core            18.1.1
@angular-devkit/schematics      18.1.1
@angular/build                  18.0.0
@schematics/angular             18.1.1
ng-packagr                      18.1.0
rxjs                            7.8.1
typescript                      5.4.5

**Comment by garrettld** [comment]
This has kind of been fixed in v19. I've created a StackBlitz here that demonstrates that it works fine in v19: https://stackblitz.com/edit/stackblitz-starters-wyf4yc?file=src%2Fmain.ts

The reason it fails in v18 is this line here: https://github.com/angular/angular/blob/main/packages/core/primitives/signals/src/watch.ts#L75-L77

You can see that the cleanup function is overwritten, which is why only the last-registered cleanup function is called. In v19 the effect implementations push cleanup functions into an array and call all of them.

The offending line of code is still present in v19, but I think it's only still around as an internal backwards compatibility thing for Google. Until/unless there's a fix in v18 or you can upgrade to v19, you should be able to work around this by putting all your cleanup logic into a single call to onCleanup.

**Comment by sheikalthaf** [comment]
@garrettld yes it is not working in V18 and working in V19 because of effect rework done in this [PR 56501](https://github.com/angular/angular/pull/56501/files#diff-d350f6dfdd77593b52a63e68ecac002c1271baf04a60dfe9ac61d1a635ce29c2) Ln 255.

Thanks for the detailed insights on this issue.

**Comment by alxhub** [comment]
+1 - this was fixed in #56501 which is landing in v19.
